File: backendfastapi/utils/websocket_manager.py
from typing import Dict, Set
from fastapi import WebSocket
import asyncio
import json
import os
import redis
from threading import Thread

class WebSocketManager:
    """ตัวจัดการ mapping session_id -> set[WebSocket]"""

    def __init__(self) -> None:
        self.active_connections: Dict[str, Set[WebSocket]] = {}
        self.redis_client = redis.Redis.from_url(os.getenv("REDIS_URL", "redis://localhost:6379/0"))
        self.pubsub = self.redis_client.pubsub()
        self.running = False
        self._start_redis_listener()

    def _start_redis_listener(self):
        """Start Redis listener in background thread"""
        if not self.running:
            self.running = True
            thread = Thread(target=self._redis_listener_worker, daemon=True)
            thread.start()

    def _redis_listener_worker(self):
        """Background worker to listen for Redis messages"""
        try:
            print("🔍 Starting Redis listener for chat channels...")
            # Subscribe to all chat channels
            self.pubsub.psubscribe("chat:*")
            print("✅ Redis listener subscribed to chat:* channels")
            
            for message in self.pubsub.listen():
                if message["type"] == "pmessage":
                    channel = message["channel"].decode()
                    session_id = channel.replace("chat:", "")
                    data = message["data"].decode()
                    
                    print(f"📨 Received Redis message for session {session_id}: {data[:100]}...")
                    
                    # Broadcast to all WebSocket connections for this session
                    asyncio.create_task(self._broadcast_to_session(session_id, data))
        except Exception as e:
            print(f"❌ Redis listener error: {e}")
        finally:
            self.running = False

    async def _broadcast_to_session(self, session_id: str, message: str):
        """Broadcast message to all WebSocket connections for a session"""
        if session_id not in self.active_connections:
            print(f"⚠️ No active connections for session {session_id}")
            return
        
        print(f"📤 Broadcasting to {len(self.active_connections[session_id])} connections for session {session_id}")
        dead_sockets = []
        for ws in list(self.active_connections[session_id]):
            try:
                await ws.send_text(message)
                print(f"✅ Message sent to WebSocket for session {session_id}")
            except Exception as e:
                print(f"❌ Failed to send to WebSocket: {e}")
                dead_sockets.append(ws)
        
        # Cleanup dead connections
        for ws in dead_sockets:
            self.disconnect(session_id, ws)

    async def connect(self, session_id: str, websocket: WebSocket):
        if session_id not in self.active_connections:
            self.active_connections[session_id] = set()
        self.active_connections[session_id].add(websocket)
        print(f"WebSocket connected to session {session_id}, total connections: {len(self.active_connections[session_id])}")

    def disconnect(self, session_id: str, websocket: WebSocket):
        if session_id in self.active_connections:
            self.active_connections[session_id].discard(websocket)
            if not self.active_connections[session_id]:
                del self.active_connections[session_id]
            print(f"WebSocket disconnected from session {session_id}")

    async def broadcast(self, session_id: str, message: str):
        if session_id not in self.active_connections:
            return
        dead_sockets = []
        for ws in list(self.active_connections[session_id]):
            try:
                await ws.send_text(message)
            except Exception:
                dead_sockets.append(ws)
        # cleanup
        for ws in dead_sockets:
            self.disconnect(session_id, ws)


ws_manager = WebSocketManager() 

File: backend/src/utils/websocketManager.ts
import { WebSocket } from 'ws';
import { redis as sharedRedis } from '../lib/redis';
import { EventEmitter } from 'events';

interface WebSocketConnection {
  ws: WebSocket;
  userId: string;
  sessionId?: string;
  isAlive: boolean;
}

export class WebSocketManager extends EventEmitter {
  private connections: Map<string, WebSocketConnection> = new Map();
  private sessionConnections: Map<string, Set<string>> = new Map();
  private redisClient: any;
  private redisSubscriber: any;
  private isRedisConnected = false;

  constructor() {
    super();
    this.initializeRedis();
  }

  private async initializeRedis() {
    try {
      // Reuse shared ioredis instance for publish; create a lightweight subscriber for pattern subscribe
      this.redisClient = sharedRedis; // ioredis instance
      this.redisSubscriber = sharedRedis.duplicate();

      // Subscribe pattern and handle messages via 'pmessage' event (ioredis semantics)
      await this.redisSubscriber.psubscribe('chat:*');
      this.redisSubscriber.on('pmessage', (_pattern: string, channel: string, message: string) => {
        try {
          const sessionId = typeof channel === 'string' ? channel.replace('chat:', '') : '';
          if (sessionId) {
            this.broadcastToSession(sessionId, message);
          }
        } catch (err) {
          console.error('❌ Error handling pmessage:', err);
        }
      });

      this.isRedisConnected = true;
      console.log('✅ Redis connected and subscribed to chat channels');
    } catch (error) {
      console.error('❌ Failed to connect to Redis:', error);
      this.isRedisConnected = false;
    }
  }

  public addConnection(connectionId: string, ws: WebSocket, userId: string): void {
    const connection: WebSocketConnection = {
      ws,
      userId,
      isAlive: true
    };

    this.connections.set(connectionId, connection);

    // Set up ping/pong for connection health
    ws.on('pong', () => {
      const conn = this.connections.get(connectionId);
      if (conn) {
        conn.isAlive = true;
      }
    });

    // Handle connection close
    ws.on('close', () => {
      this.removeConnection(connectionId);
    });

    console.log(`✅ WebSocket connected: ${connectionId} for user: ${userId}`);
  }

  public joinSession(connectionId: string, sessionId: string): void {
    const connection = this.connections.get(connectionId);
    if (!connection) {
      console.warn(`⚠️ Connection ${connectionId} not found`);
      return;
    }

    connection.sessionId = sessionId;

    if (!this.sessionConnections.has(sessionId)) {
      this.sessionConnections.set(sessionId, new Set());
    }
    this.sessionConnections.get(sessionId)!.add(connectionId);

    console.log(`✅ User ${connection.userId} joined session ${sessionId}`);
  }

  // Legacy method for backward compatibility
  public connect(sessionId: string, ws: WebSocket): void {
    // This method is for legacy compatibility
    // In the new implementation, we use connectionId-based approach
    console.log(`🔗 Legacy connect called for session ${sessionId}`);
  }

  public leaveSession(connectionId: string): void {
    const connection = this.connections.get(connectionId);
    if (!connection || !connection.sessionId) {
      return;
    }

    const sessionId = connection.sessionId;
    const sessionConnections = this.sessionConnections.get(sessionId);
    if (sessionConnections) {
      sessionConnections.delete(connectionId);
      if (sessionConnections.size === 0) {
        this.sessionConnections.delete(sessionId);
      }
    }

    connection.sessionId = undefined;
    console.log(`👋 User ${connection.userId} left session ${sessionId}`);
  }

  public removeConnection(connectionId: string): void {
    const connection = this.connections.get(connectionId);
    if (!connection) {
      return;
    }

    // Leave session if connected to one
    if (connection.sessionId) {
      this.leaveSession(connectionId);
    }

    this.connections.delete(connectionId);
    console.log(`❌ WebSocket disconnected: ${connectionId}`);
  }

  public sendToConnection(connectionId: string, message: string): boolean {
    const connection = this.connections.get(connectionId);
    if (!connection) {
      return false;
    }

    try {
      connection.ws.send(message);
      return true;
    } catch (error) {
      console.error(`❌ Failed to send message to ${connectionId}:`, error);
      this.removeConnection(connectionId);
      return false;
    }
  }

  public broadcastToSession(sessionId: string, message: string): void {
    const sessionConnections = this.sessionConnections.get(sessionId);
    if (!sessionConnections) {
      console.log(`⚠️ No connections for session ${sessionId}`);
      return;
    }

    const deadConnections: string[] = [];
    let sentCount = 0;

    for (const connectionId of sessionConnections) {
      if (this.sendToConnection(connectionId, message)) {
        sentCount++;
      } else {
        deadConnections.push(connectionId);
      }
    }

    // Clean up dead connections
    deadConnections.forEach(connectionId => {
      this.removeConnection(connectionId);
    });

    console.log(`📤 Broadcasted to ${sentCount} connections in session ${sessionId}`);
  }

  // Legacy method for backward compatibility
  public broadcast(sessionId: string, message: string): void {
    this.broadcastToSession(sessionId, message);
  }

  public broadcastToUser(userId: string, message: string): void {
    let sentCount = 0;
    const deadConnections: string[] = [];

    for (const [connectionId, connection] of this.connections) {
      if (connection.userId === userId) {
        if (this.sendToConnection(connectionId, message)) {
          sentCount++;
        } else {
          deadConnections.push(connectionId);
        }
      }
    }

    // Clean up dead connections
    deadConnections.forEach(connectionId => {
      this.removeConnection(connectionId);
    });

    console.log(`📤 Broadcasted to ${sentCount} connections for user ${userId}`);
  }

  public publishToRedis(channel: string, message: string): void {
    if (!this.isRedisConnected || !this.redisClient) {
      console.warn('⚠️ Redis not connected, cannot publish message');
      return;
    }

    this.redisClient.publish(channel, message)
      .catch((error: any) => {
        console.error(`❌ Failed to publish to Redis:`, error);
      });
  }

  public getConnectionCount(): number {
    return this.connections.size;
  }

  public getSessionConnectionCount(sessionId: string): number {
    const sessionConnections = this.sessionConnections.get(sessionId);
    return sessionConnections ? sessionConnections.size : 0;
  }

  public getConnectionInfo(connectionId: string): WebSocketConnection | undefined {
    return this.connections.get(connectionId);
  }

  // Health check for connections
  public pingConnections(): void {
    for (const [connectionId, connection] of this.connections) {
      if (!connection.isAlive) {
        console.log(`💀 Removing dead connection: ${connectionId}`);
        this.removeConnection(connectionId);
        continue;
      }

      connection.isAlive = false;
      try {
        connection.ws.ping();
      } catch (error) {
        console.error(`❌ Failed to ping connection ${connectionId}:`, error);
        this.removeConnection(connectionId);
      }
    }
  }

  // Cleanup method
  public async cleanup(): Promise<void> {
    // Close all WebSocket connections
    for (const [connectionId, connection] of this.connections) {
      try {
        connection.ws.close();
      } catch (error) {
        console.error(`❌ Error closing connection ${connectionId}:`, error);
      }
    }

    this.connections.clear();
    this.sessionConnections.clear();

    // Close Redis connections
    if (this.redisClient) {
      await this.redisClient.quit();
    }
    if (this.redisSubscriber) {
      await this.redisSubscriber.quit();
    }

    console.log('🧹 WebSocket manager cleaned up');
  }
}

// Export singleton instance
export const wsManager = new WebSocketManager(); 